### PR TITLE
Features/vllm async test

### PR DESCRIPTION
**Pull request by p-ferreira**
## NOTE: INCLUDES PR #165

- adds `threaded_log` and `asyng_log` utility functions
- evolves async implementation from PR #118

### CODE REVIEW

**Review by schampoux — APPROVED** [review]

File: tests/test_forward.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+def generate_reference(x, delay=1):
+    time.sleep(delay)
+    return 'Fake reference'
+
+@pytest.mark.parametrize('delay', [0.1, 0.2, 0.3])
+@pytest.mark.parametrize('timeout', [0.1, 0.2])
+@pytest.mark.parametrize('min_time', [0, 0.05, 0.1])
+@pytest.mark.parametrize('max_time', [0.1, 0.15, 0.2])
+def test_generate_reference_while_waiting_for_dendrite(delay, timeout, min_time, max_time):
```
Comment (steffencruz) [inline tests/test_forward.py:27]: Any recommended improvements or enhancements to these tests? Are they still measuring what matters?

File: prompting/forward.py [hunk 2]
```diff
@@ -54,17 +68,23 @@ async def run_step(
...
-    )
 
+    reference_generation_task = None
+    with ThreadPoolExecutor() as executor:
+        # Start the task to generate the reference
+        if not agent.task.static_reference:            
+            reference_generation_task = executor.submit(generate_reference, agent)        
+            
+        # Prepare the tasks
+        responses = await execute_dendrite_call(self.dendrite(axons=axons, synapse=PromptingSynapse(roles=["user"], messages=[agent.challenge]), timeout=timeout))
```
[the remark this answers is not in the document]
Comment (p-ferreira) [reply prompting/forward.py]: Exactly, the reference will run in parallel to the network call

File: prompting/forward.py [hunk 3]
```diff
@@ -94,13 +114,12 @@ async def run_step(
         **response_event.__state_dict__(),
     }
 
-    log_event(self, event)
```
[the remark this answers is not in the document]
Comment (p-ferreira) [reply prompting/forward.py:97]: Yeah, I had to do this to get the forward_time metric and that was the easiest acceptable solution from my pov

**Review by steffencruz — APPROVED** [review]
